File: src/Levitate.hs
{-# LANGUAGE ForeignFunctionInterface #-}
module Levitate where

{-@ assume (+↓) :: a:{v:Float | true} -> b:{v:Float | true} -> {v:Float | v = a +↓ b} @-}
foreign import ccall "add_down" (+↓) :: Float -> Float -> Float

{-@ assume (+↑) :: a:{v:Float | true} -> b:{v:Float | true} -> {v:Float | v = a +↑ b} @-}
foreign import ccall "add_up"   (+↑) :: Float -> Float -> Float

{-@ assume (-↓) :: a:{v:Float | true} -> b:{v:Float | true} -> {v:Float | v = a -↓ b} @-}
foreign import ccall "sub_down" (-↓) :: Float -> Float -> Float

{-@ assume (-↑) :: a:{v:Float | true} -> b:{v:Float | true} -> {v:Float | v = a -↑ b} @-}
foreign import ccall "sub_up"   (-↑) :: Float -> Float -> Float

{-@ assume (*↓) :: a:{v:Float | true} -> b:{v:Float | true} -> {v:Float | v = a *↓ b} @-}
foreign import ccall "mul_down" (*↓) :: Float -> Float -> Float

{-@ assume (*↑) :: a:{v:Float | true} -> b:{v:Float | true} -> {v:Float | v = a *↑ b} @-}
foreign import ccall "mul_up"   (*↑) :: Float -> Float -> Float

{-@ assume (/↓) :: a:{v:Float | true} -> b:{v:Float | true} -> {v:Float | v = a /↓ b} @-}
foreign import ccall "div_down" (/↓) :: Float -> Float -> Float

{-@ assume (/↑) :: a:{v:Float | true} -> b:{v:Float | true} -> {v:Float | v = a /↑ b} @-}
foreign import ccall "div_up"   (/↑) :: Float -> Float -> Float

-- (.=)

-- use this for all runtime floating-point comparisons instead of (==)
-- do not pattern match on float literals (which is equivalent to using ==)

{-@ assume (.=) :: a:{v:Float | true} -> b:{v:Float | true} -> {v:Bool | (Prop v) <=> a .= b}  @-}
foreign import ccall "ieee_equality" (.=) :: Float -> Float -> Bool


{-@ pinf :: {v:Float | v = (pinf 0)}  @-}
pinf :: Float
pinf = (1.0)/(0.0)

{-@ ninf :: {v:Float | v = (ninf 0)} @-}
ninf :: Float
ninf = -pinf

{-@ pzero :: {v:Float | v = (pzero 0)} @-}
pzero :: Float
pzero = 0.0

{-@ nzero :: {v:Float | v = (nzero 0)} @-}
nzero :: Float
nzero = 1.0/ninf

{-@ nonNan :: x:Float -> {v:Bool | (Prop v) <=> not (isNan x)} @-}
nonNan :: Float -> Bool
nonNan x = (x .= x) 

File: src/levitate.c
#include "fenv.h"

// return the rounded-down result of adding two arguments,
// where the real value of the sum of the arguments is rounded down
// to the nearest lower floating-point number.
// 
// cleans up after itself by setting the rounding mode to RoundToNearest,
// which is the default mode supported by haskell
float add_down(float a, float b)
{
  float ret;
  fesetround(FE_DOWNWARD);
  ret = a + b;
  fesetround(FE_TONEAREST);
  return ret;
}


// return the rounded-up result of adding two arguments,
// where the real value of the sum of the arguments is rounded up
// to the nearest higher floating-point number.
// 
// cleans up after itself by setting the rounding mode to RoundToNearest,
// which is the default mode supported by haskell
float add_up(float a, float b)
{
  float ret;
  fesetround(FE_UPWARD);
  ret = a + b;
  fesetround(FE_TONEAREST);
  return ret;
}


// return the rounded-down result of subtracting two arguments,
// where the real value of the difference of the arguments is rounded down
// to the nearest lower floating-point number.
// 
// cleans up after itself by setting the rounding mode to RoundToNearest,
// which is the default mode supported by haskell
float sub_down(float a, float b)
{
  float ret;
  fesetround(FE_DOWNWARD);
  ret = a - b;
  fesetround(FE_TONEAREST);
  return ret;
}


// return the rounded-up result of subtracting two arguments,
// where the real value of the difference of the arguments is rounded up
// to the nearest higher floating-point number.
// 
// cleans up after itself by setting the rounding mode to RoundToNearest,
// which is the default mode supported by haskell
float sub_up(float a, float b)
{
  float ret;
  fesetround(FE_UPWARD);
  ret = a - b;
  fesetround(FE_TONEAREST);
  return ret;
}

// return the rounded-down result of multiplying two arguments,
// where the real value of the product of the arguments is rounded down
// to the nearest lower floating-point number.
// 
// cleans up after itself by setting the rounding mode to RoundToNearest,
// which is the default mode supported by haskell
float mul_down(float a, float b)
{
  float ret;
  fesetround(FE_DOWNWARD);
  ret = a * b;
  fesetround(FE_TONEAREST);
  return ret;
}


// return the rounded-up result of adding two arguments,
// where the real value of the product of the arguments is rounded up
// to the nearest higher floating-point number.
// 
// cleans up after itself by setting the rounding mode to RoundToNearest,
// which is the default mode supported by haskell
float mul_up(float a, float b)
{
  float ret;
  fesetround(FE_UPWARD);
  ret = a * b;
  fesetround(FE_TONEAREST);
  return ret;
}


// return the rounded-down result of dividing two arguments,
// where the real value of the quotient of the arguments is rounded down
// to the nearest lower floating-point number.
// 
// cleans up after itself by setting the rounding mode to RoundToNearest,
// which is the default mode supported by haskell
float div_down(float a, float b)
{
  float ret;
  fesetround(FE_DOWNWARD);
  ret = a / b;
  fesetround(FE_TONEAREST);
  return ret;
}


// return the rounded-up result of dividing two arguments,
// where the real value of the quotient of the arguments is rounded up
// to the nearest higher floating-point number.
// 
// cleans up after itself by setting the rounding mode to RoundToNearest,
// which is the default mode supported by haskell
float div_up(float a, float b)
{
  float ret;
  fesetround(FE_UPWARD);
  ret = a / b;
  fesetround(FE_TONEAREST);
  return ret;
}


// returns true whenever the two arguments are equal
// this exists to workaround issues with liquid haskell:
// we need both ieee-equality (for haskell's builtin == operator) 
// and bitwise equality (so that we can distinguish between +0 and -0)
// so the hack workaround is to make bitwise equality the default in liquidhaskell refinements,
// and implement a special function to perform ieee-equality.  
int ieee_equality(float a, float b)
{
  return a == b;
}
